daemon/config/config.go

```diff
@@ -687,6 +721,12 @@ func Validate(config *Config) error {
 		}
 	}
 
+	if len(config.Features) > 0 {
+		if err := validateFeatures(config.Features); err != nil {
+			return err
+		}
+	}
```

Comment:
This is a (minor) breaking change, right?  It's only breaking for really nutty users, and I don't know how common my "hack" is, but one of my ["spacebar heater"](https://xkcd.com/1172/) behaviors is using the `features` object as a free-form comments field because every other part of this JSON file is now strictly validated (so I personally expected this validation to happen in the API endpoint).  For example, here's a real actual snippet from one of my production configuration files:

```json
	"features": {
		"000 - keys of the features object is the only place we can sneak comments that the engine won't complain about": false,
		"userland-proxy: in *theory*, we can set userland-proxy to false and avoid the userland proxy binary; in practice, that makes it so containers in non-default networks cannot access 172.17.0.1 for DNS / squignix (and neither configuration gives us access to external host IPs correctly, like redacted-external-hostname)": false
	},
```

(perhaps this is better suited as a feature request for something like `Comments json.RawMessage` or `Ignored` as a new field :disappointed: :heart:)

With this change, I don't think there's anywhere left we can put "comments" of any kind within this configuration file. :sob:

Reply:
For sure! I did wonder about this — and thought "maybe it's fine, there's **literally** no point to having any other extra keys right now so maybe it's okay.

I never anticipated you'd be using it for comments 😅 (where's your god now, JSON-folk?)

In the other PR I had this validation on the endpoint and we discussed it a bit and with @thaJeztah's input ended up removing it in favor of validating it on load, but I'm happy to remove this even.

Another option would be to ignore values that don't conform while loading (instead of erroring out).

Reply:
> I never anticipated you'd be using it for comments 😅 (where's your god now, JSON-folk?)

/me cries in https://nigeltao.github.io/blog/2021/json-with-commas-comments.html

> Another option would be to ignore values that don't conform while loading (instead of erroring out).

I like this one; honestly we have a fixed set we support, so we could go further and have a central list where we filter "anything else" explicitly, but I don't feel too strongly. I'm mostly just lamenting the last place I can hide "comments" and hope we can find a compromise that let's us keep comments in the config in some form. ❤️

One option for that would be to stop marshalling it as a `map` and instead make an explicit struct of booleans, but we might've made our parser validate _that_ too, IIRC.

Reply:
> One option for that would be to stop marshalling it as a `map` and instead make an explicit struct of booleans, but we might've made our parser validate that too, IIRC.

Right, I considered that too – however, from some of the convos with other maintainers, I think an incidental benefit of this change is that it makes for an easy way for custom runtimes/providers (like DD, etc.) to advertise custom features that specific clients can understand. Being able to just add custom values in the `daemon.json` is a benefit there, so I'm okay with allowing "well-formatted" unknown values. That said, it's not essential, so I'm okay making that change if we decide it's not it.

I'll change it to ignore non-conforming values for now.

Reply:
LOL, I was a bit behind reading up on comments, but TIANON!!! What hackery are you putting on us??

Did you consider using daemon labels?

```json
{
	"features": {
		"containerd-snapshotter": true
	},
	"labels": [
		"xyz.tianon.comment.000=keys of the features object is the only place we can sneak comments that the engine won’t complain about",
    		"xyz.tianon.comment.001=in *theory*, we can set userland-proxy to false and avoid the userland proxy binary; in practice, that makes it so containers in non-default networks cannot access 172.17.0.1 for DNS / squignix (and neither configuration gives us access to external host IPs correctly, like redacted-external-hostname)"
	]
}
```


```bash
docker info --format '{{json .Labels}}' | jq .
[
  "xyz.tianon.comment.000=keys of the features object is the only place we can sneak comments that the engine won’t complain about",
  "xyz.tianon.comment.001=in *theory*, we can set userland-proxy to false and avoid the userland proxy binary; in practice, that makes it so containers in non-default networks cannot access 172.17.0.1 for DNS / squignix (and neither configuration gives us access to external host IPs correctly, like redacted-external-hostname)"
]
```

Reply:
Considering what options we have here;

- overall, I think having a validation for the format is, well, good?
- but, indeed, we didn't validate before this, so question is how much we care about spacebar heating
- For the `/info` _endpoint_ I think we should be ok to already be strict;
    - invalid _formatted_ options won't be returned (sorry!)
    - we _can_ return unknown, but valid formatted options (AKA "custom options")
    - :point_up: one slight thing I'm considering here is whether we should support unknown options as part of the _daemon configuration_
    - :point_up: my initial thinking about allowing the list to be "extensible" was for the _API_ (i.e., if Docker or Mirantis proxies the API, they can append options to the response)
    - :point_right: this would allow the daemon to still validate and/or warn about options it doesn't know about
    - :point_right: which can be a great time-saver for users that may have set a `feature` and made a typo (`--feature containerd-snapshooter` would produce a warning or error)
    - :bulb: :point_right: we _COULD_ (given "conventions" we discussed and to mark non-namespaced options as "reserved") allow unknown, but namespaced options to be set
        - i.e.: `xyz.tianon.tianon-custom-feature: true` unknown, but namespaced, so :white_check_mark:
        - but `containerd-snapshooter: true` unknown, and not namespaced, so could be a mistake, so :x: or :warning:
- We can explicitly deprecate use of non-namespaced, invalid features, but for the first release make it a warning ("DEPRECATED: xxxxx")
- And release after: "FATAL: invalid daemon config"